### ISSUE TITLE
Having an issue with Dagger Hilt Proguard Rules

### ISSUE THREAD

**Issue by Dzikry123** [google/dagger#4259]
I'm a beginner Android developer facing a challenge with Proguard while using Dagger Hilt for dependency injection in my project. Despite trying various solutions, I'm still encountering errors related to missing classes generated by Dagger Hilt.

In essence, when attempting to build my app, I receive errors indicating missing classes from Dagger Hilt, even after implementing Proguard rules suggested in the Dagger repository. This issue has proven to be quite stubborn, and I'm seeking guidance from the community to resolve it.

Could someone experienced with Dagger Hilt and Proguard offer insights into properly configuring Proguard to address these errors? here is my code in my project:

### **// proguard-rules.pro :**

  # Keep file R and Manifest
    -keep class **.R
    -keep class **.R$* {*;}
    -keep class **.BuildConfig {*;}
    -keep class **.Manifest {*;}

    -keep class androidx.constraintlayout.motion.widget.KeyAttributes { *; }

    # Keep Dependency Injection Framework related classes and methods
    -keep class dagger.hilt.** { *; }
    -keep class javax.inject.** { *; }
    -keep class javax.annotation.** { *; }

    # Keep Model and Repository Classes
    -keep class com.example.core.data.model.** { *; }
    -keep class com.example.core.data.repo.** { *; }

    # Keep ViewModels
    -keep class * extends androidx.lifecycle.ViewModel

    # Keep Parcelable Classes
    -keep class * implements android.os.Parcelable {
        public static final android.os.Parcelable$Creator *;
    }

    # Keep UI-related classes
    -keep class com.example.ui.activity.** { *; }
    -keep class com.example.ui.fragment.** { *; }
    -keep class com.example.service.** { *; }

    # Menjaga semua kelas dalam paket tertentu
    -keep class com.example.package.** { *; }

    # Menjaga semua kelas yang disebutkan dalam file tertentu
    #-keep @com.example.rules.txt

    # Menjaga kelas model
    -keep class com.example.model.** { *; }

    # Menjaga kelas service
    -keep class com.example.service.** { *; }

    # Menjaga kelas-kelas yang diperlukan untuk penanganan SSL/TLS
    -keep class okhttp3.internal.platform.ConscryptPlatform {*;}
    -keep class okhttp3.internal.platform.OpenJSSEPlatform {*;}
    -keep class org.bouncycastle.** {*;}
    -keep class org.conscrypt.** {*;}
    -keep class org.openjsse.** {*;}

    # Keep classes required by OpenJSSE
    -keep class sun.security.x509.** { *; }
    -keep class sun.util.logging.** { *; }

    -keep class com.example.projectone.GithubUserApplication { *; }
    -keep class com.example.projectone.ui.main.MainActivity { *; }
    -keep class com.example.projectone.ui.userdetail.DetailUserActivity { *; }

# Keep classes generated by Hilt
    -keep class com.example.projectone.*Hilt* {
        *;
    }
# Keep Hilt generated classes
```
-keep class com.example.projectone.Hilt_* {*;}
-keep,allowobfuscation,allowshrinking @dagger.hilt.EntryPoint class *
-keep,allowobfuscation,allowshrinking @dagger.hilt.android.EarlyEntryPoint class *

```

#missing_rules.txt

```
-dontwarn com.example.projectone.Hilt_GithubUserApplication
-dontwarn com.example.projectone.ui.main.Hilt_MainActivity
-dontwarn com.example.projectone.ui.userdetail.Hilt_DetailUserActivity
```

here is the error : 

> Missing class com.example.projectone.Hilt_GithubUserApplication (referenced from: void com.example.projectone.GithubUserApplication.() and 1 other context)

> Missing classes detected while running R8. Please add the missing classes or apply additional keep rules that are generated in missing_rules.txt (see the file above)
 
> Missing class com.example.projectone.ui.main.Hilt_MainActivity (referenced from: void com.example.projectone.ui.main.MainActivity.() and 5 other contexts)
 
> Missing class com.example.projectone.ui.userdetail.Hilt_DetailUserActivity (referenced from: void com.example.projectone.ui.userdetail.DetailUserActivity.() and 5 other contexts)

> Caused by: [CIRCULAR REFERENCE: com.android.tools.r8.utils.b: Missing class com.example.projectone.Hilt_GithubUserApplication (referenced from: void com.example.projectone.GithubUserApplication.() and 1 other context)

Any assistance or advice would be greatly appreciated. Thank you for your time and your support.

**Comment by Dzikry123** [comment]
i have try to follow the [official guidance](https://developer.android.com/build/shrink-code#enable) and put all of the rules into my proguard-rules.pro, and i have include the `getDefaultProguardFile("proguard-android-optimize.txt").` too in my build.gradle, but somehow the error is still exist, i put the link for my project repository below so you can see the detail of my code 

[My Repository](https://github.com/Dzikry123/Android-Proguard/blob/main/core/build.gradle)

**Comment by wanyingd1996** [comment]
Hi, I patched your repro and ran `./gradlew assembleDebug`, but I'm not seeing the missing dagger errors, I only see missing StringConcatFactory, which looks like https://github.com/Kotlin/kotlinx.serialization/issues/2145. But I noticed that your repo is using pretty old version of  Dagger, Kotlin, Ksp and Compose. Can you try to upgrade your versions first to see if the errors goes away?

**Comment by Dzikry123** [comment]
Hello, i have tried to update all of the version of my Dagger, Kotlin, Ksp and Compose but it seems the error is remains the same,

![Screenshot 2024-03-15 141931](https://github.com/google/dagger/assets/96764302/966b0665-ab44-4ae3-841e-0ea270a48147)
![Screenshot 2024-03-15 142012](https://github.com/google/dagger/assets/96764302/f09f315b-9fff-4c56-a58b-f88bc921df8c)


 i have updated my repository and here is the link to the updated repository :
[Repository-Update](https://github.com/Dzikry123/Android-Proguard/tree/version2)

**Comment by wanyingd1996** [comment]
Hi, I noticed that the hilt gradle plugin is missing from your class path. https://github.com/Dzikry123/Android-Proguard/blob/version2/build.gradle#L9.  https://dagger.dev/hilt/gradle-setup.html, please refer to the hilt gradle section for how to include it in the class path.

**Comment by aldee** [comment]
@wanyingd1996 he's already declared it in https://github.com/Dzikry123/Android-Proguard/blob/b0c85c6d5ba802f6190c5237f4dbd74a482b7089/build.gradle#L17 using the new plugin DSL.

I encountered similar issue as well after upgrade AGP to 8.3 and Gradle 8.6, which uses R8 full mode by default.

**Comment by aldee** [comment]
@wanyingd1996 no worries. actually I don't have that problem anymore, somehow. my problem was actually on runtime, but it's complaining about similar issue

`Fatal Exception: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/example/app/MyApplication_GeneratedInjector;
       at com.example.DaggerMyApplication_HiltComponents_SingletonC.builder(DaggerMyApplication_HiltComponents_SingletonC.java:64)
       at com.example.Hilt_MyApplication$1.get(Hilt_MyApplication.java:24)
       at dagger.hilt.android.internal.managers.ApplicationComponentManager.generatedComponent(ApplicationComponentManager.java:40)
       at com.example.Hilt_MyApplication.generatedComponent(Hilt_MyApplication.java:37)
       at com.example.Hilt_MyApplication.hiltInternalInject(Hilt_MyApplication.java:52)
       at com.example.Hilt_MyApplication.onCreate(Hilt_MyApplication.java:43)
       at com.example.MyApplication.onCreate(MyApplication.kt:13)
       at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1316)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6998)
       at android.app.ActivityThread.-$$Nest$mhandleBindApplication()
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2236)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:205)
       at android.os.Looper.loop(Looper.java:294)
       at android.app.ActivityThread.main(ActivityThread.java:8177)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)`

that's the stacktrace of the crash, and that crash only happened after using R8 full mode. But unfortunately I forgot what the environment was (gradle, hilt, AGP version).

Maybe @Dzikry123 can try updating to latest AGP and Gradle? and also use the latest R8 as well by adding this

`classpath("com.android.tools:r8:8.3.37")` to your project's build.gradle buildscript dependencies

**Comment by Dzikry123** [comment]
Hi @aldee and @wanyingd1996 thank you for your support, i have tried to update the AGP, Gradle and also the R8 to my project's build.gradle , but the error is still exist, i try to update the android studio to iguana version, but the error remains the same, here is the update of my project, perhaps you can find some clue here :

[Repo-update-2](https://github.com/Dzikry123/Android-Proguard/tree/version2)

**Comment by aldee** [comment]
So in my case, I'm using Hilt version 2.44.2, it works fine. But when I upgraded it to 2.51, I got the same error with nothing else changed.

**Comment by wanyingd1996** [comment]
Hi, @Dzikry123, there are several misconfiguration in the project.
(1) `:favorite` and `:core` are libraries, you don't need to include the android default proguard files there. I'm seeing warning about it is not necessary to include those in a module.
(2) You can using incompatible version of hilt-android and hilt-compiler, these are separate artifact, but using different version of them may result in internal API incompatible.
(3) You did not include hilt as dependency for `:core`, but you are defined hilt entry point in that library, therefore, you have a missing binding error hidden behind the r8 error.

Applying the above changes resulted in a successful build from my side. I think the missing class from r8 is just red herrings, covering the real errors.

To summarize, It is recommended to alway try the latest version of libraries to avoid seeing mysterious errors, and some of the bugs may already being fixed in the latest version.

**Comment by Dzikry123** [comment]
Hi @wanyingd1996 and @aldee after 2 weeks i found what is the error, the warnings I see (e.g., "Class com.example.projectone.Hilt_GithubUserApplication does not exist") are caused by these classes not being present in the input passed to the compiler.

i tried to get Hilt to work by updating the version and minifyEnabled false then tried switching to minifyEnabled true once it worked.

thank you so much for helping me to find the solution of this problem :)

**Comment by bahacan19** [comment]
@Dzikry123 Hi,
I pulled your repository and am getting same error as you did before. is there any certain ways to get rid of it? I already tired switching off and on again minifyEnabled flag. No change...
